llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/VectorExt/IR/VectorExtAttrs.td

```diff
@@ -76,11 +76,20 @@ def LayoutAttr : IREEVectorExt_Attr<"Layout",
   let assemblyFormat = "`<`$layouts`>`";
   let genVerifyDecl = 0;
   let extraClassDeclaration = [{
+    // Get the shape for a given layout dimension.
+    std::optional<int64_t> getShape(const LayoutDimension &dim);
     std::optional<int64_t> getBatchDim(int64_t dim);
+    // Get the lane dimension shape for a provided simd tensor dim.
+    std::optional<int64_t> getLaneDim(int64_t dim);
+    // Get the lane dimension for a provided simd tensor dim.
+    std::optional<LayoutDimension> getLane(int64_t dim);
 
     // Returns the grid of lane ids. Assumes a valid layout.
     ::std::tuple<int64_t, int64_t, int64_t> getLaneGrid();
     PerDimLayoutAttr getDimLayout(int64_t dim) const;
+    // Given the reduction dim, computes a shuffle offset
+    // based on the shapes of the lane dimensions.
+    uint64_t getShuffleOffset(int64_t reductionDim);
```

Comment:
I don't think this should be part of the layout defination.

Reply:
Why not? The shuffle offset that is being computed is very specific to the layout.

Reply:
IMO this is fine to add here, but it would be good to try to differentiate between strategy specific helpers and more general stuff.

I found the naming a little confusing though, maybe call it `getResidualThreadCount`? Or clarify in the comment that it is computing the number of threads involved in the reduction (correct me if that's not what it's doing).

Reply:
Makes sense. So let me try to explain what this is doing here. So when we do parallel reductions using butterfly shuffles (this happens inside doThreadGlobalReduction), we need to specify the shuffle offset for the gpuShuffleOp. The value that's returned by the getShuffleOffset function is not the number of threads involved in the reduction. It is the starting offset for the butterfly shuffle. So say shuffleOffset was 1 and laneSize was 2, that would mean that there are 2 threads involved in the reduction. The laneSize parameter contains the number of threads involved in the reduction (and this is directly readable from the layout since this maps to either LANEX, LANEY or LANEZ). The shuffleOffset is the offset required for the gpu Shuffle Op which is specified as ```For lane k returns the value from lane k ^ offset```. Let me know if this clarifies things. I will add more comments to explain this as well.

Reply:
Ah I see, makes more sense looking at the docs for `gpu.shuffle` too. Thanks for the clarification. Yeah comments would definitely help to clarify the connection to the name.

Still pretty much indifferent to where this helper goes though, but it is more problem specific than I thought.